lib/util/zip.ex

```diff
@@ -0,0 +1,143 @@
+# Copyright(c) 2015-2019 ACCESS CO., LTD. All rights reserved.
+
+use Croma
+
+defmodule Antikythera.Zip do
+  @moduledoc """
+  Wrapper module for `zip` command.
+
+  For the consistency in working with antikythera and other gears, scope of this module is limited under a temporary directory reserved by `Antikythera.Tmpdir.make/2`.
+
+  Functions only accept absolute paths for both source and resulting archive.
+  """
+
+  alias Croma.Result, as: R
+  alias Antikythera.Context
+  alias Antikythera.ExecutorPool.Id, as: EPoolId
+  alias AntikytheraCore.TmpdirTracker
+
+  @typep opts :: {:encryption, boolean} | {:password, String.t}
+
+  defmodule FileName do
+    use Croma.SubtypeOfString, pattern: ~r/^(?!.*\/\.{0,2}\z).*\z/
+  end
+
+  @doc """
+  Creates a ZIP file.
+
+  Encryption using a password is supported in which `encryption` option is set `true`.
+
+  ## Example
+    Tmpdir.make(context, fn tmpdir ->
+      src_path = tmpdir <> "/src.txt"
+      zip_path = tmpdir <> "/archive.zip"
+      File.write!(src_path, "text")
+      Antikythera.Zip.zip(context, zip_path, src_path, [encryption: true, password: "password"])
+      |> case do
+        {:ok, archive_path} ->
+          ...
+      end
+      ...
+    end)
+  """
+  defun zip(
+    context_or_epool_id :: v[EPoolId.t | Context.t],
+    zip_raw_path        :: v[FileName.t],
+    src_raw_path        :: v[String.t],
+    opts                :: v[list(opts)] \\ []
+  ) :: R.t(Path.t) do
+    epool_id = extract_epool_id(context_or_epool_id)
+    with(
+      {:ok, tmpdir} <- TmpdirTracker.get(epool_id),
+      :ok           <- reject_existing_dir(zip_raw_path),
+      :ok           <- validate_suffix(src_raw_path),
+      zip_path      <- Path.expand(zip_raw_path),
+      src_path      <- Path.expand(src_raw_path),
+      :ok           <- validate_within_tmpdir(zip_path, tmpdir),
+      :ok           <- validate_within_tmpdir(src_path, tmpdir),
+      :ok           <- validate_path_exists(src_path),
+      :ok           <- ensure_dir_exists(zip_path, tmpdir),
+      {:ok, args}   <- opts |> Map.new() |> extract_zip_args(),
+      :ok           <- try_zip_cmd(args ++ [zip_path, src_path])
+    ) do
```

Comment:
For the same reason as https://github.com/access-company/antikythera/pull/44#discussion_r288398820:
```
with {:ok, tmpdir} <- TmpdirTracker.get(epool_id),
     ...,
     ... do
```

However, this comment is optional because I think it is easy to read even in the current writing style.

Reply:
Fixed in 2a164a4 .

> it is easy to read even in the current writing style.

It could be, but it also depends on how to make decisions about coding style.